File: server/src/main/com/broll/gainea/server/core/bot/impl/BotPlace.java
package com.broll.gainea.server.core.bot.impl;

import com.broll.gainea.net.NT_Action_PlaceUnit;
import com.broll.gainea.net.NT_PlayerTurnActions;
import com.broll.gainea.net.NT_Reaction;
import com.broll.gainea.server.core.actions.required.PlaceUnitAction;
import com.broll.gainea.server.core.bot.BotAction;
import com.broll.gainea.server.core.bot.BotUtils;
import com.broll.gainea.server.core.map.Location;

import java.util.List;

public class BotPlace extends BotAction<NT_Action_PlaceUnit> {
    @Override
    protected void handleAction(NT_Action_PlaceUnit action, NT_Reaction reaction) {
        if (action.possibleLocations.length == 1) {
            reaction.option = 0;
            return;
        }
        PlaceUnitAction handler = game.getReactionHandler().getActionHandlers().getHandler(PlaceUnitAction.class);
        List<Location> locations = BotUtils.getLocations(game, action.possibleLocations);
        int index = strategy.chooseUnitPlace(handler.getUnitToPlace(), locations);
        reaction.option = index;
    }

    @Override
    public Class<NT_Action_PlaceUnit> getActionClass() {
        return NT_Action_PlaceUnit.class;
    }

    @Override
    public float score(NT_Action_PlaceUnit action, NT_PlayerTurnActions turn) {
        return -1;
    }
}

File: server/src/main/com/broll/gainea/server/core/bot/impl/BotPlace.kt
package com.broll.gainea.server.core.bot.impl

import com.broll.gainea.net.NT_Action
import com.broll.gainea.net.NT_Action_PlaceUnit
import com.broll.gainea.net.NT_Reaction
import com.broll.gainea.server.core.actions.required.PlaceUnitAction
import com.broll.gainea.server.core.bot.BotAction
import com.broll.gainea.server.core.bot.BotUtils

class BotPlace : BotAction<NT_Action_PlaceUnit>() {
    override fun react(action: NT_Action_PlaceUnit, reaction: NT_Reaction) {
        if (action.possibleLocations.size == 1) {
            reaction.option = 0
            return
        }
        val handler = game.reactionHandler.actionHandlers.getHandler(PlaceUnitAction::class.java)
        val locations = BotUtils.getLocations(game, action.possibleLocations)
        val index = strategy.chooseUnitPlace(handler.unitToPlace!!, locations)
        reaction.option = index
    }

    override val actionClass: Class<out NT_Action>
        get() = NT_Action_PlaceUnit::class.java
}
